### DISCUSSION TITLE
Bounds / Box / Range types

### DISCUSSION THREAD

**Discussion by mcclure** [bitshifter/glam-rs#375] [General]
I situation I frequently find myself in when working with vectors is I have a 2D rectangle or 3D volume and I need to know if a particular coordinate is "in bounds", IE, I have an axis-aligned bounding box and I need to know if a point is within it. It is straightforward to keep a minimum and maximum coordinate and >=/<= each component, but this is verbose and therefore error-prone. I like to have a single object encapsulating the bounds especially since this makes it easy to tack on convenience features such as "expand the bounds to include this point" and "offset the bounds".

There doesn't seem to be a good library for this with glam or in Rust in general. "Glamour" has Box2D/Box3D structs, but Glamour is somewhat heaviweight and expects you to use its own special specially-typed vector struct idiom, also it currently doesn't seem to be compatible with Glam 0.22 (requires 0.21). It would be great to have a Box/Bounds/AABB/multidimensional range struct that uses Glam types natively.

I previously contributed [Bound2 and Bound3](https://github.com/excessive/cpml/tree/master/modules) classes to the CPML vector math library for Lua, and would be happy to create a PR or addon crate for Rust implementing a bounds object. It would be helpful though if while I work on the PR I could get feedback on how to fit glam's idioms best (For starters: should it be named Box2, Bound2 or something else? Would Box<IVec2> be preferred or specialized IBox2/DBox3 type classes? Would it be better to do this as a glam-rs PR or a new crate?).

**Comment by bitshifter** [comment]
I think for now this might be better as a separate crate just in that I'm not really keeping up with `glam` maintenance at the moment and don't want to add to the pile. As for naming, I'm not sure. `Box2`/`Box3` would fit with glam naming I suppose. Bevy has a `Rect` already. There are other crates like Embark Studio's macaw that has a 3D `BoundingBox` among other things.

Probably he main question / difficulty is if more than `f32` support is wanted. Probably there will be users who do want that. Then you might want to use the same code generation stuff that glam uses, which complicates things somewhat. I could probably split that out into its own project if it was needed. Also `Vec3` vs `Vec3A`. Most people just use `Vec3` but `Vec3A` usually performs better.

**Comment by mcclure** [comment]
Thanks for the feedback.

I may just do a Bounds<T> that parameterizes on a glam type. If I'm working separately from glam then that makes it less weird to break glam style and use a generic…

(Unfortunately generic Box<T> is out for obvious reasons…)

It you think this is totally out feel free to close.

**Comment by bitshifter** [comment]
Converted to a discussion. I think this is a useful thing to do, but not something I want to take on myself right now. This might be a good place to co-ordinate if someone decides to take it on though.
